mm2src/common/custom_futures/repeatable.rs

```diff
@@ -0,0 +1,785 @@
+
+impl<F, T, E> RepeatableTrait<T, E> for F where F: Future<Output = Action<T, E>> + Unpin {}
+
+pub(crate) trait InspectErrorTrait<E>: 'static + Fn(&E) + Send {}
+
+impl<F: 'static + Fn(&E) + Send, E> InspectErrorTrait<E> for F {}
+
+#[derive(Clone, Debug, PartialEq)]
+pub enum RepeatError<E> {
+    TimeoutExpired {
+        until: Until,
+        /// An error occurred during the last attempt.
+        error: E,
+    },
+    AttemptsExceed {
+        attempts: usize,
+        /// An error occurred during the last attempt.
+        error: E,
+    },
+}
+
+impl<E> RepeatError<E> {
+    pub fn error(&self) -> &E {
+        match self {
+            RepeatError::TimeoutExpired { error, .. } | RepeatError::AttemptsExceed { error, .. } => error,
+        }
+    }
+
+    pub fn into_error(self) -> E {
+        match self {
+            RepeatError::TimeoutExpired { error, .. } | RepeatError::AttemptsExceed { error, .. } => error,
+        }
+    }
+
+    fn timeout(until: Until, error: E) -> Self { RepeatError::TimeoutExpired { until, error } }
+
+    fn attempts(attempts: usize, error: E) -> Self { RepeatError::AttemptsExceed { attempts, error } }
+}
+
+impl<E: fmt::Display> fmt::Display for RepeatError<E> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            RepeatError::TimeoutExpired { until, error } => {
+                write!(
+                    f,
+                    "Waited too long until {until:?} for the future to succeed. Error: {error}",
+                )
+            },
+            RepeatError::AttemptsExceed { attempts, error } => {
+                write!(f, "Error {error} on retrying the future after {attempts} attempts")
+            },
+        }
+    }
+}
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum Until {
+    Instant(Instant),
+    TimestampMs(u64),
+}
+
+/// The future is ether ready (with a `T` result), or not ready (failed with an intermediate `E` error).
+#[derive(Debug)]
+pub enum Action<T, E> {
+    Ready(T),
+    Retry(E),
+}
+
+pub trait RetryOnError<T, E> {
+    fn retry_on_err(self) -> Action<T, E>;
+}
+
+impl<T, E> RetryOnError<T, E> for Result<T, E> {
+    /// Converts `Result<T, E>` into `Action<T, E>`:
+    /// * `Ok(T)` => `Action::Ready(T)`.
+    /// * `Err(E)` => `Action::Retry(E)`.
+    #[inline]
+    fn retry_on_err(self) -> Action<T, E> {
+        match self {
+            Ok(ready) => Action::Ready(ready),
+            Err(e) => Action::Retry(e),
+        }
+    }
+}
+
+/// The result of `repeatable` or `retry_on_err` macros - the first step at the future configuration.
+pub struct Repeatable<Factory, F, T, E> {
+    factory: Factory,
+    /// Currently executable future, i.e. an active attempt.
+    exec_fut: Option<F>,
+    /// A timeout future if we're currently waiting for a timeout.
+    timeout_fut: Option<Timer>,
+    until: Option<Until>,
+    attempts: Option<AttemptsState>,
+    repeat_every: Option<Duration>,
```

Comment:
I think `repeat_every` might have a default value instead of being `Option`.

Reply:
Do you mean `Duration::default()` that equals to `0s`?

Reply:
Done

Reply:
`0s` is dangerous - if `repeat_every` is not explicitly set, the future will be repeated without any delay, causing excessive resource usage. I would prefer it to be 1 second at least.

Reply:
Makes sense, will do!

Reply:
Done